### DISCUSSION TITLE
USB RNDIS functionality on STM32U5 board

### DISCUSSION THREAD

**Discussion by enthunilu** [zephyrproject-rtos/zephyr#51765] [General] [answered]
Hi All,

I am trying to build sample code [](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/net/sockets/dumb_http_server)
 to test USB-RNDIS functionality on nucleo_u575zi_q board. Same code is working for nucleo_f429zi board.

When using same code for nucleo_u575zi_q board, device is getting detected as RNIDS Compatible Device. However on console I am getting below errors.
"Not enough pkt buffers, len 134, skip 70 " 
"Skip 64 bytes out of remaining 70 bytes"
"Skip 6 bytes out of remaining 6 bytes"
"Not enough pkt buffers, len 140, skip 76 " 
"Skip 64 bytes out of remaining 76 bytes"

This error is coming from https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device/class/netusb/function_rndis.c

`			LOG_ERR("Not enough pkt buffers, len %u, skip %u",
				rndis.in_pkt_len, rndis.skip_bytes);`

It looks like number of bytes are around 140 and I am running USB in full speed with 64 bytes of size. But same is true for nucleo_f429zi board as well. In which I dont get such error.

Then I started debugging it, however could not find anything yet. I also found while building code, I am getting warning as 
"No SOURCES given to Zephyr library: drivers__ethernet"

![image](https://user-images.githubusercontent.com/93653544/198646877-6e0e818f-76ed-460e-afbb-1d3da2705452.png)

Because of this error, started wondering whether we need ethernet driver to make usb-rndis work or there is some other issue with this.
Since, STM32U5 controller does not have ethernet support.

Can anyone please share some thoughts on this ?

Thanks for your kind support!!

**Comment by FRASTM** [comment]
Did you try with b_u585i_iot02a platform ? 
There is a usb otg device in its dts (not seen in the nucleo_u575)

**Reply by enthunilu** [reply]
Hi @FRASTM,

Unfortunately I do not have b_u585i_iot02a board. So, I tried to check usb_device node property available for b_u585i_iot02a board and added same to nucleo_u575zi board overlay file.

zephyr_udc0: &usbotg_fs {
	pinctrl-0 = <&usb_otg_fs_dm_pa11 &usb_otg_fs_dp_pa12>;
	pinctrl-names = "default";
	status = "okay";
};
I might have missed to add other required configs, so I started debugging one by one to understand what all things I need to consider for the same. But could not get it clearly.

**Reply by enthunilu** [reply]
If some pointer or something you can suggest to look at that will help a lot.

**Reply by Masood-Salik** [reply]
@enthunilu May you tell me how you made the code work for nucleo_f429zi board? 
[Problem Description](https://github.com/zephyrproject-rtos/zephyr/discussions/69056)

**Comment by FRASTM** [comment]
I built the samples/net/zperf/  and for any stm32u5 target board, What I see on the console is 
```

*** Booting Zephyr OS build zephyr-v3.2.0-1148-g9a9fc2b8a5dd ***
[00:00:00.000,000] <inf> usb_rndis: Command thread started
[00:00:00.000,000] <inf> usb_net: netusb initialized
[00:00:00.010,000] <inf> net_config: Initializing network
[00:00:00.010,000] <inf> net_config: Waiting interface 1 (0x20000b98) to be up...
[00:00:30.011,000] <err> net_config: Timeout while waiting network interface
[00:00:30.011,000] <err> net_config: Network initialization failed (-115)
uart:~$
```

with prj.conf including :
CONFIG_USB_DEVICE_NETWORK_RNDIS=y
CONFIG_USB_DEVICE_STACK=y
CONFIG_USB_DEVICE_NETWORK_ECM=n

**Reply by enthunilu** [reply]
Yeah actually to get device get detected as RNIDS device I also did some change in usb descriptor as mentioned in below discussion
https://github.com/zephyrproject-rtos/zephyr/discussions/47865
Like this,
![image](https://user-images.githubusercontent.com/93653544/199650169-6b28bbdf-af9a-43ac-b4e0-1c4357df90b7.png)

After that I was able to get IP address allocated and device got detected as RNIDS device.

**Reply by GeorgeCGV** [reply]
That approach works to enable auto device detection on WIndows 10 Version 21H2.

Moving https://github.com/zephyrproject-rtos/zephyr/issues/46268#issuecomment-1344394010 to this discussion.

> > I close it because Zephyr's RNDIS Class/SubClass/Protocol codes are already the same as described in https://learn.microsoft.com/en-us/windows-hardware/drivers/network/interface-descriptor-for-communication-class-interface.
> 
>  interesting as https://learn.microsoft.com/en-us/azure/rtos/usbx/usbx-device-stack-supplemental-2 descriptors differ from https://learn.microsoft.com/en-us/windows-hardware/drivers/network/interface-descriptor-for-communication-class-interface.
> 
> However, with 
> ```
> .bFunctionClass = USB_BCC_MISCELLANEOUS,
> .bFunctionSubClass = 4, 
> .bFunctionProtocol = 1,
> ```
> and 
> ```
> CONFIG_USB_DEVICE_NETWORK_ECM=n
> CONFIG_USB_DEVICE_NETWORK_RNDIS=y
> CONFIG_USB_COMPOSITE_DEVICE=y
> ```
> everything works out of the box.
> 
> `CONFIG_USB_COMPOSITE_DEVICE=y` solves another error related to the device descriptor when IAD is used.

**Comment by enthunilu** [comment]
Hello @FRASTM  did you get chance to look at this further ?

**Reply by FRASTM** [reply]
The [function_rndis.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device/class/netusb/function_rndis.c) 
 already defines 
`USBD_CLASS_DESCR_DEFINE(primary, 0) struct usb_rndis_config rndis_cfg = {`

In case the parameters like _bFunctionClass_ or _bFunctionSubClass_ or _bFunctionProtocol_
have to be changed, maybe through a DTS property or so, this is more a question of `subsys/usb/`  code
@jfischer-no, any advice ?

**Reply by jfischer-no** [reply]
> The [function_rndis.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/device/class/netusb/function_rndis.c) already defines `USBD_CLASS_DESCR_DEFINE(primary, 0) struct usb_rndis_config rndis_cfg = {`
> 
> In case the parameters like _bFunctionClass_ or _bFunctionSubClass_ or _bFunctionProtocol_ have to be changed, maybe through a DTS property or so, this is more a question of `subsys/usb/` code @jfischer-no, any advice ?

DTS must not be used to configure "_bFunctionClass_ or _bFunctionSubClass_ or _bFunctionProtocol_ ". That also not something to configure. It is either right or it is not.

**Reply by FRASTM** [reply]
Ok but still to clear to me what to do with that because I  am not familiar with the USB
-->    why  a change is needed is this a bug a new configuration, What is wrong ?
Why should the interface descriptor in the rndis function  be changed with setting corresponding to the "Base Class EFh (Miscellaneous) " and subclass = 0x04h  protocol = 0x1 --> "RNDIS over Ethernet" 
Is the stm32U5 not compatible with this USB class or what is missing to make it support that ?

**Reply by enthunilu** [reply]
Thanks @FRASTM  and @jfischer-no for your inputs.
These are something new points for me as a learning part.

Actually, if we use existing function_rndis.c class descriptor as it is, device does not get detected as RNDIS device. ( Looks like a bug )
To make it work, I did changes as mentioned above got idea from usb.org ( https://www.usb.org/defined-class-codes )

With these changes I tested on nucleo_f429zi board and it worked, I mean I was able to see board is getting detected as RNDIS Device and also i was able to ping the device and was able to load default web page.

Same when I tried to do with nucleo_u575zi_q board i started getting error as mentioned at the start of the ticket.

**Reply by jfischer-no** [reply]
This cannot be specific to STM32U5 or other platforms.  I will look at it closely today and probably push a patch.

**Reply by FRASTM** [reply]
> This cannot be specific to STM32U5 or other platforms. I will look at it closely today and probably push a patch.

thanks @jfischer-no for your support

**Reply by jfischer-no** [reply]
It is not clear to me what the "RNDIS over Ethernet" values from https://www.usb.org/defined-class-codes refers to, it does not look right because Zephyr's RNDIS is a CDC device with two interfaces and has to be `bInterfaceClass = USB_BCC_CDC_CONTROL`.
According to Microsoft RNDIS documentation
https://learn.microsoft.com/en-us/windows-hardware/drivers/network/interface-descriptor-for-communication-class-interface, we actually have right values for the interface descriptors, maybe association descriptor SubClass and Protocol codes should be changed to be inline with codes from the first interface.

**Reply by jfischer-no** [reply]
@enthunilu Can you please just enable Kconfig option CONFIG_USB_COMPOSITE_DEVICE (that changes Class/SubClass/Protocol codes in the device descriptors, https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/usb-interface-association-descriptor) and retest?

**Reply by enthunilu** [reply]
Thanks @jfischer-no, I'll check and update here.

**Reply by enthunilu** [reply]
Hello @jfischer-no,

Sorry for the delayed response. I was facing some setup related issue due to which I was not able to build the code.

I tried what you suggested and found that, device does not get detected as RNDIS device. We need to make changes in descriptor.
with below change,
`CONFIG_USB_COMPOSITE_DEVICE=y`
`CONFIG_USB_DEVICE_NETWORK_RNDIS=y`

Interestingly, just to give it a try, I took latest release ( Zephyr 3.2.0 and zephyr-sdk-0.15.2 ) and tried my same code without any changes in function_rndis.c and by enabling 
`CONFIG_USB_COMPOSITE_DEVICE=y`
`CONFIG_USB_DEVICE_NETWORK_RNDIS=y`

my nucleo_u575zi_q board is getting detected as RNDIS Device and also it is working fine. I am not getting any issue as mentioned at the start.

Thank you so much @jfischer-no  @FRASTM  @GeorgeCGV for all your support.

Earlier I was using Zephyr 3.1.0 since I have a requirement of webserver and CivetWeb is deprecated in Zephyr 3.2.0 due to which I was not doing upgrade.

Just to make sure this test, I'll try same test on one more system. And update here.

Thanks Again!! :-)

**Reply by enthunilu** [reply]
I reverified USB-RNDIS functionality on nucleo_u575zi_q board and it is working fine with latest Zephyr version.
No change is required at code side. Only project config changes are enough to make it work.

Thank you so much @FRASTM @jfischer-no @GeorgeCGV for all your support.

**Reply by GeorgeCGV** [reply]
@enthunilu, tested with v3.3.0-rc1 and without code modification USB device is not detected as RNDIS automatically. Driver has to be manually selected for the device in the device manager. However, Windows detects the device as RNDIS automatically with mentioned code modifications. (_STM32H7_)

**Reply by ritumohan** [reply]
@enthunilu Could you please list me the project configuration which is required to test Civetweb over RNDIS. Tested Civetweb over Ethernet, need support to test civetweb over RNDIS

**Comment by enthunilu** [comment, marked as answer]
With zephyr-v3.2.0 and zephyr-sdk-0.15.2 and config setting suggested by @jfischer-no  and @FRASTM 
USB-RNDIS functionality works fine.

**Reply by GeorgeCGV** [reply]
Based on https://learn.microsoft.com/en-us/windows-hardware/drivers/usbcon/supported-usb-classes proposed USB descriptor modification is required to make Windows use RNDIS driver automatically:

> Class: Miscellaneous (EFh), SubClass: 04h, Protocol: 01h
